File: ruby/lib/runapi/wan/resources/edit_video.rb
# frozen_string_literal: true

module RunApi
  module Wan
    module Resources
      # Modifies existing videos guided by a text prompt and optional reference image.
      # The 2.6 models use source_video_urls (plural, required) while 2.7 uses
      # source_video_url (singular, required). Flash variants support audio and multi-shot.
      class EditVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/wan/edit_video"
        RESPONSE_CLASS = Types::VideoTaskResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedVideoTaskResponse

        def initialize(http)
          @http = http
        end

        # Edit a video and wait until complete.
        #
        # @param params [Hash] video editing parameters
        # @return [RunApi::Wan::Types::CompletedVideoTaskResponse] completed video edit
        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        # Create a video editing task.
        #
        # @param params [Hash] video editing parameters
        # @return [RunApi::Wan::Types::VideoTaskResponse] task creation result with id
        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        # Get video editing status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Wan::Types::VideoTaskResponse] current task status
        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["edit-video"], params)
        end
      end
    end
  end
end

File: python/src/runapi/wan/resources/edit_video.py
"""Wan edit-video resource."""

from __future__ import annotations

from typing import Any, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import (
    CompletedVideoTaskResponse,
    VideoTaskResponse,
)


class EditVideo(Resource):
    """Edit videos with Wan models."""

    ENDPOINT = "/api/v1/wan/edit_video"

    RESPONSE_CLASS = VideoTaskResponse
    COMPLETED_RESPONSE_CLASS = CompletedVideoTaskResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Edit a video and poll until it completes.

        Args:
            **params: video edit parameters (model, ...).

        Returns:
            The completed (narrowed) video edit response.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a video edit task and return immediately with an id.

        Args:
            **params: video edit parameters (model, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["edit-video"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a video edit task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)
